rclcpp/multithread_executor/multithread_executor.cpp

```diff
@@ -0,0 +1,188 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <chrono>
+#include <condition_variable>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <thread>
+#include <iostream>
+
+#include "rclcpp/rclcpp.hpp"
+#include "std_msgs/msg/string.hpp"
+
+using namespace std::chrono_literals;
+
+/* For this example, we will be creating a publishing node like the one in minimal_publisher.
+ * We will have a single subscriber node running 2 threads. Each thread loops at different speeds, and
+ * just repeats what it sees from the publisher to the screen.
+ */
+
+class PublisherNode : public rclcpp::Node
+{
+public:
+  PublisherNode()
+  : Node("PublisherNode")
+  {
+    publisher_ = this->create_publisher<std_msgs::msg::String>("topic", 10);
+    auto timer_callback =
+      [this]() -> void {
+        auto message = std_msgs::msg::String();
+        message.data = "Hello World! " + std::to_string(this->count_++);
+
+        // Extract current thread
+        auto curr_thread = string_thread_id(std::this_thread::get_id());
+
+        // Prep display message
+        auto info_message = "\n<<THREAD " + curr_thread + ">> Publishing '%s'";
+        RCLCPP_INFO(this->get_logger(), info_message, message.data.c_str());
+        this->publisher_->publish(message);
+      };
+    timer_ = this->create_wall_timer(500ms, timer_callback);
+  }
+
+private:
+  /**
+   * A small convenience function for converting a thread ID to a string
+   */
+  std::string string_thread_id(std::thread::id id)
```

Comment:
Two things about this method:

1.  It could be a function at the top of the file; that way you can share it between the Publisher and Subscriber.
1.  It could just call `std::this_thread::get_id()` itself, which would simplify the interface to it.

Reply:
This comment is still open.

Reply:
Whoops, sorry about that